src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -687,23 +690,48 @@ ModifyQuerySupported(Query *queryTree, Query *originalQuery, bool multiShardQuer
 	}
 
 	/*
-	 * Reject queries which involve joins. Note that UPSERTs are exceptional for this case.
-	 * Queries like "INSERT INTO table_name ON CONFLICT DO UPDATE (col) SET other_col = ''"
-	 * contains two range table entries, and we have to allow them.
+	 * We have to allow modify queries with two range table entries, if it is pushdownable.
 	 */
 	if (commandType != CMD_INSERT && queryTableCount != 1)
```

Comment:
As discussed on slack, the `queryTableCount != 1` does not belong here because we can have subqueries without having multiple tables. It would make sense to move it into the `if` statement below, since it does apply to `master_modify_multiple_shards`